Changed usages of hash algo md5 to xxh128

tinect opened the pull request:
### 1. Why is this change necessary?
Performance
see #4778

### 2. What does this change do, exactly?
- removed the useless re-hashing in [NotFoundSubscriber::generateKey](https://github.com/shopware/shopware/compare/trunk...tinect:feat/NEXT-38371?expand=1#diff-7c60bcc3f674ac4aa62dd0752b2b97bd03f5b4012e15326dc8cde0bf064387a5) of ContextHash
- Changed usages of hash algo md5 to xxh128 in several places
- Kernel::getCacheHash: (total half of time spent)
  - changed algo to xxh128: -32% time spent
  - removed the hashing of plugin data while it is being hashed again and removed the substr: more -50%
- Changed the hashed in the tests

### 3. Describe each step to reproduce the issue or behaviour.


### 4. Please link to the relevant issues (if any).
Closes #4778
Closes #4774

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

tinect committed:
feat: added dedicated hasher and updated usages of md5 hashing

ghost commented:
@tinect The `'xxh128'` could perhaps be moved to a central public constant. This would make it more consistent to use and make future changes to a different hash algorithm easier since only one change would be necessary.

tinect commented:
> @tinect The `'xxh128'` could perhaps be moved to a central public constant. This would make it more consistent to use and make future changes to a different hash algorithm easier since only one change would be necessary.

I like the idea. I'd like to see, what shopware is thinking about it.

One point could be that there might be problems coming in the future, which we might not know for now. So, I think it's better that it is being checked usage by usage, not generally.

shyim commented on tests/unit/Core/Framework/DataAbstractionLayer/Cache/EntityCacheKeyGeneratorTest.php:
```diff
@@ -50,44 +50,40 @@ public function testBuildStreamTag(): void
 
-        static::assertSame($hash, $generator->getCriteriaHash($criteria));
+        static::assertNotEmpty($hash);
+        static::assertIsString($hash);
```
this makes the test useless. of course the hash is an string, but we don't know if it does differ now.
that's why there was hard-code hashes to check that there happens a difference

tinect replied:
I've reverted the change and updated the hashes in tests

keulinho commented:
Hey @tinect , 

thanks for your contribution! 🎉 

While looking at the changes here and in the other two MRs we wondered if it is really necessary that we hard code the hashing algo all over the place, what do you think about extracting it, so we can easily replace that under the hood if something faster is available?

Additional benefit would be that it is probably easier to get the community to adopt that wrapper as well, so their extensions benefit from the perf boost as well? 

WDYT? We could probably accompany that change with a custom phprule that prohibits use of md5 and hashing function directly and thus "enforce" that also for all new cases we would use the wrapper to we don't end up with mutliple suboptimal hashing algos being used all over the place

tinect commented:
> Hey @tinect ,
> 
> thanks for your contribution! 🎉
> 
> While looking at the changes here and in the other two MRs we wondered if it is really necessary that we hard code the hashing algo all over the place, what do you think about extracting it, so we can easily replace that under the hood if something faster is available?
> 
> Additional benefit would be that it is probably easier to get the community to adopt that wrapper as well, so their extensions benefit from the perf boost as well?
> 
> WDYT? We could probably accompany that change with a custom phprule that prohibits use of md5 and hashing function directly and thus "enforce" that also for all new cases we would use the wrapper to we don't end up with mutliple suboptimal hashing algos being used all over the place

Great, thank you.
Uhhh yeah. I like having a general util class. I added this one. As a constant, I added the current used algo, so anyone can save this information, if it is necessary - f.e. reproducible hashes. WDYT? :-)

What do you think of accepting mixed type in the `hash` function and implementing `Json::encode` if the given data is no string?

I really like the idea of the phprule, can you point me to the direction, where I have to look? Otherwise, are you at the hacktoberfest - so we could finalize this thing if you like?!

I split the PRs due to several topics they touch. But maybe it should be merged into each other by now?

keulinho commented:
Yes i will be at the hacktoberfest, so latest we can work on that there together, could also be a follow up.

You can find the existing PhpStan rules we have in /src/Core/DevOps/StaticAnalyze/Phpstan

keulinho commented:
Hey @tinect, we'ved merged your PR, only made some smaller changes (e.g. decoupling the Util method names and signature from the underlying implementation etc)

Thx again for your contribution!
